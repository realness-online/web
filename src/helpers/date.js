import { as_created_at } from '@/helpers/itemid'
const this_year = Date.parse(new Date().getFullYear())
const format_as_time = {
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}
const format_as_day = {
  weekday: 'long',
  day: 'numeric',
  month: 'long'
}
const format_as_day_and_year = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  year: 'numeric'
}
const format_as_day_and_time = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}
export default {
  as_time (date, format = format_as_time) {
    const time = new Date(date)
    return time.toLocaleString('en-US', format) // TODO: get country code from browser
  },
  as_day (date) {
    let day
    if (Date.parse(date) < this_year) {
      day = this.as_time(date, format_as_day_and_year)
    } else day = this.as_time(date, format_as_day)
    if (this.is_same_day(date, new Date())) day = 'Today'
    return day
  },
  as_day_and_time (date) {
    return this.as_time(date, format_as_day_and_time)
  },
  is_same_day (d1, d2) {
    d1 = new Date(d1)
    d2 = new Date(d2)
    return d1.getFullYear() === d2.getFullYear() &&
      d1.getDate() === d2.getDate() && d1.getMonth() === d2.getMonth()
  },
  day_name (date) {
    date = new Date(date)
    return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`
  },
  id_as_day (itemid) {
    return new Date(as_created_at(itemid))
  }

}
