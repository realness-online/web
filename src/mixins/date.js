import date_helper from '@/helpers/date'
export default {
  data () {
    return {
      today: this.as_day(new Date().toISOString())
    }
  },
  methods: {
    as_day (iso_string) {
      return date_helper.as_day(iso_string)
    },
    as_time (iso_string) {
      return date_helper.as_time(iso_string)
    },
    as_day_and_time (iso_string) {
      return date_helper.as_day_and_time(iso_string)
    }
  }
}
