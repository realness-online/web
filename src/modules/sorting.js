export default {
  newer_first(earlier, later) {
    return Date.parse(later.created_at) - Date.parse(earlier.created_at)
  },
  older_first(earlier, later) {
    return Date.parse(earlier.created_at) - Date.parse(later.created_at)
  }
}
