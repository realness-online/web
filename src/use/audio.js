import { ref } from 'vue'

/**
 * @typedef {Object} AudioAnalyzer
 * @property {boolean} analyzing - Whether audio analysis is currently active
 * @property {(stream: MediaStream) => void} analyze_audio - Start analyzing audio from stream
 */

/**
 * Audio analysis composable
 * @returns {AudioAnalyzer}
 */
export const use = () => {
  /** @type {import('vue').Ref<AnalyserNode|null>} */
  const analyser = ref(null)

  /** @type {import('vue').Ref<boolean>} */
  const analyzing = ref(false)

  /**
   * Processes audio data on each animation frame
   */
  const process_audio = () => {
    const times = new Uint8Array(analyser.value.frequencyBinCount)
    analyser.value?.getByteTimeDomainData(times)
    // for (var i = 0; i < times.length; i++) {
    //   const value = times[i]
    //   const zerod = value - 128
    //   console.log(i, value, zerod)
    // }
    if (analyzing.value) requestAnimationFrame(process_audio)
  }

  /**
   * Initializes audio analysis for a media stream
   * @param {MediaStream} stream - The audio stream to analyze
   */
  const analyze_audio = stream => {
    const context = new AudioContext()
    analyser.value = context.createAnalyser()
    analyser.value.fftSize = 32
    analyser.value.minDecibels = -64
    analyser.value.maxDecibels = -6
    const input = context.createMediaStreamSource(stream)
    input.connect(analyser.value)
    analyzing.value = true
    requestAnimationFrame(process_audio)
  }

  return {
    analyzing,
    analyze_audio
  }
}
