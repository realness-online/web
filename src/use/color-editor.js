export function change_color(id, color, type = 'fill') {
  const path = this.get_path_element(id)
  path.setAttribute(type, color)
}

export function get_color(id, type = 'fill') {
  return this.get_path_element(id).getAttribute(`${type}`)
}

export function get_path_element(id) {
  return document.getElementById(id).querySelector('path')
}

export function get_path_attribute(id, type = 'fill') {
  return this.get_path_element(id).getAttribute(`${type}`)
}
