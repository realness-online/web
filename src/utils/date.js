import { as_created_at } from '@/utils/itemid'
const this_year = Date.parse(new Date().getFullYear())
const format_as_time = {
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}
const format_as_day = {
  weekday: 'long',
  day: 'numeric',
  month: 'long'
}
const format_as_day_and_year = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  year: 'numeric'
}
const format_as_day_and_time = {
  weekday: 'long',
  day: 'numeric',
  month: 'long',
  hour: 'numeric',
  minute: 'numeric',
  hour12: true
}

export function today_as_date() {
  const now = new Date()
  return `${now.getMonth() + 1}/${now.getDate()}/${now.getFullYear()}`
}

export function is_today(a_date) {
  if (a_date === today_as_date()) return true
  return false
}
export function is_same_day(d1, d2) {
  d1 = new Date(d1)
  d2 = new Date(d2)
  return (
    d1.getFullYear() === d2.getFullYear() &&
    d1.getDate() === d2.getDate() &&
    d1.getMonth() === d2.getMonth()
  )
}
export function is_fresh(date = 0) {
  const expires = new Date()
  expires.setDate(expires.getDate() - 13)
  if (new Date(date) > expires) return true
  return false
}
export function as_time(date, format = format_as_time) {
  const time = new Date(date)
  return time.toLocaleString('en-US', format) // TODO: get country code from browser
}
export function as_day(date) {
  let day
  if (Date.parse(date) < this_year) day = as_time(date, format_as_day_and_year)
  else day = as_time(date, format_as_day)
  if (is_same_day(date, new Date())) day = 'Today'
  return day
}
export function as_day_and_time(date) {
  return as_time(date, format_as_day_and_time)
}
export function day_name(date) {
  date = new Date(date)
  return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`
}
export function id_as_day(itemid) {
  return day_name(as_created_at(itemid))
}
export const format_time_remaining = time_ms => {
  const minutes = Math.floor(time_ms / 60000)
  const hours = Math.floor(minutes / 60)
  const remaining_minutes = minutes % 60

  if (hours > 0) return `${hours}h ${remaining_minutes}m`

  return `${minutes}m`
}
