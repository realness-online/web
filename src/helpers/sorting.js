import { as_created_at } from '@/helpers/itemid'
export function recent_id_first (first, second) {
  return as_created_at(second) - as_created_at(first)
}
export function recent_item_first (first, second) {
  return as_created_at(second.id) - as_created_at(first.id)
}
export function recent_visit_first (first, second) {
  return new Date(first.visited) - as_created_at(second.visited)
}

export function recent_date_first (first, second) {
  return new Date(second[0]) - new Date(first[0]) // newer is larger
}
export function recent_number_first (first, second) {
  return parseInt(second) - parseInt(first)
}
export function earlier_weirdo_first (first, second) {
  return recent_id_first(get_id(second), get_id(first))
}
export function recent_weirdo_first (first, second) {
  return recent_id_first(get_id(first), get_id(second))
}
function get_id (thing) {
  if (Array.isArray(thing)) return thing[0].id
  else return thing.id
}
