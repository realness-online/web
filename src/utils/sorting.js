/** @typedef {import('@/types').Id} Id */
/** @typedef {import('@/types').Item} Item */
/** @typedef {import('@/types').Type} Type */
/** @typedef {import('@/types').Person} Person */

import { as_created_at } from '@/utils/itemid'

/**
 * @param {any} thing
 * @returns {Id}
 */
export const get_id = thing => {
  if (Array.isArray(thing)) return thing[0].id
  return thing.id
}

/**
 * @param {Id|null} first
 * @param {Id|null} second
 * @returns {number}
 */
export const recent_id_first = (first, second) =>
  as_created_at(second) - as_created_at(first)

/**
 * @param {Item} first
 * @param {Item} second
 * @returns {number}
 */
export const recent_item_first = (first, second) =>
  as_created_at(second.id) - as_created_at(first.id)

/**
 * @param {Person} first
 * @param {Person} second
 * @returns {number}
 */
export const recent_visit_first = (first, second) =>
  new Date(second?.visited).getTime() - new Date(first?.visited).getTime()

/**
 * @param {Array<string>} first
 * @param {Array<string>} second
 * @returns {number}
 */
export const recent_date_first = (first, second) =>
  new Date(second[0]).getTime() - new Date(first[0]).getTime() // newer is larger

/**
 * @param {number} first
 * @param {number} second
 * @returns {number}
 */
export const recent_number_first = (first, second) => {
  console.log('recent_number_first', first, second)
  const first_number = parseInt(String(first))
  const second_number = parseInt(String(second))

  return second_number - first_number
}

/**
 * @param {Item} first
 * @param {Item} second
 * @returns {number}
 */
export const earlier_weirdo_first = (first, second) =>
  recent_id_first(get_id(second), get_id(first))

/**
 * @param {Item} first
 * @param {Item} second
 * @returns {number}
 */
export const recent_weirdo_first = (first, second) =>
  recent_id_first(get_id(first), get_id(second))

/**
 * @param {number} time_1
 * @param {number} time_2
 * @returns {number}
 */
export const newest_timestamp_first = (time_1, time_2) => time_2 - time_1
