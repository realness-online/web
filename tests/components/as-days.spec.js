import { shallowMount } from '@vue/test-utils'
import as_days from '@/components/as-days'
import get_item from '@/use/item'
import * as as_date from '@/use/date'
import * as sorting from '@/use/sorting'

const statements_html = read_mock_file('@@/html/statements.html')
const poster_html = read_mock_file('@@/html/poster.html')
const poster = get_item(poster_html)
const statements = get_item(statements_html).statements
describe('@/components/as-days', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(as_days)
  })
  describe('Renders', () => {
    it('With nothing', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
    it('With posters and statements sorted into the days they were created', async () => {
      await wrapper.setProps({ statements, posters: [poster] })
      expect(wrapper.element).toMatchSnapshot()
      expect([...wrapper.vm.days.entries()].length).toBe(4)
    })
  })
  describe('Methods', () => {
    describe('#insert_into_day', () => {
      let other_poster
      beforeEach(() => {
        other_poster = { ...poster }
        other_poster.id = '/+16282281824/posters/559666932000'
      })
      it('Adds poster to the same day', async () => {
        vi.spyOn(sorting, 'earlier_weirdo_first')
        await wrapper.setProps({ statements, posters: [poster, other_poster] })
        expect(wrapper.vm.days.get('9/26/1987').length).toBe(2)
        expect(sorting.earlier_weirdo_first).toBeCalled()
      })
      it('Sorts todays items by most recent', async () => {
        vi.spyOn(as_date, 'is_today').mockImplementation(() => true)
        vi.spyOn(sorting, 'recent_weirdo_first')
        await wrapper.setProps({ statements, posters: [poster, other_poster] })
        expect(sorting.recent_weirdo_first).toBeCalled()
      })
    })
    describe('#check_intersection', () => {
      it('waits for day to intersect', () => {
        wrapper.vm.check_intersection([{ isIntersecting: false }])
      })
      it('Increases the number of visible days', () => {
        expect(wrapper.vm.page).toBe(1)
        wrapper.vm.days = new Set()
        for (let i = 0; i < 6; i++) {
          wrapper.vm.days.add([])
        }
        wrapper.vm.check_intersection([{ isIntersecting: true }])
        expect(wrapper.vm.page).toBe(2)
      })
      it('Paginates when there are more then 5 days', () => {
        expect(wrapper.vm.page).toBe(1)
        wrapper.vm.days = new Set()
        wrapper.vm.check_intersection([{ isIntersecting: true }])
        expect(wrapper.vm.page).toBe(1)
      })
    })
  })
})
