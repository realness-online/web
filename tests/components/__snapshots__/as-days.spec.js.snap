// Vitest Snapshot v1

exports[`@/components/as-days > Renders > With nothing 1`] = `
<section
  class="as-days"
>
  
  
</section>
`;

exports[`@/components/as-days > Renders > With posters and statements sorted into the days they were created 1`] = `
<section
  class="as-days"
>
  
  <article
    class="day"
  >
    <header>
      <h4>
        Monday, September 30, 2019
      </h4>
    </header>
    
    
    
    
  </article>
  <article
    class="day"
  >
    <header>
      <h4>
        Tuesday, September 24, 2019
      </h4>
    </header>
    
    
    
    
  </article>
  <article
    class="day"
  >
    <header>
      <h4>
        Sunday, September 22, 2019
      </h4>
    </header>
    
    
    
    
  </article>
  <article
    class="day"
  >
    <header>
      <h4>
        Saturday, September 26, 1987
      </h4>
    </header>
    
    
    
    
  </article>
  
</section>
`;
